src/Expr.cc

```diff
@@ -4653,14 +4731,20 @@ ExprPtr check_and_promote_expr(ExprPtr e, TypePtr t) {
     TypeTag t_tag = t->Tag();
 
     if ( t_tag == TYPE_ANY ) {
-        if ( e_tag != TYPE_ANY )
-            return make_intrusive<CoerceToAnyExpr>(e);
+        if ( e_tag != TYPE_ANY ) {
+            auto c = make_intrusive<CoerceToAnyExpr>(e);
+            c->SetLocation(e);
```

Comment:
I haven't checked all the call-sites, but this looks like a lot of churn easy to miss to keep consistent.

If we do this for most cases now, could we just move it into the `UnaryExpr` and `BinaryExpr` constructor to ensure it? Essentially, make new expressions use their first op's location implicitly. If in a given situation that's not appropriate, they can still do `SetLocation()` to something else.

```
diff --git a/src/Expr.cc b/src/Expr.cc
index 20d5042b6..94d687b0f 100644
--- a/src/Expr.cc
+++ b/src/Expr.cc
@@ -502,6 +502,8 @@ TraversalCode ConstExpr::Traverse(TraversalCallback* cb) const {
 UnaryExpr::UnaryExpr(ExprTag arg_tag, ExprPtr arg_op) : Expr(arg_tag), op(std::move(arg_op)) {
     if ( op->IsError() )
         SetError();
+
+    SetLocationInfo(arg_op->GetLocationInfo());
 }
 
 ValPtr UnaryExpr::Eval(Frame* f) const {
diff --git a/src/Expr.h b/src/Expr.h
index c4c60e156..443806a5f 100644
--- a/src/Expr.h
+++ b/src/Expr.h
@@ -552,6 +552,8 @@ protected:
             return;
         if ( op1->IsError() || op2->IsError() )
             SetError();
+
+        SetLocationInfo(op1->GetLocationInfo());
     }
```

Reply:
Unfortunately that change is at odds to with how the parser manages location information associated with AST nodes. It calls `set_location()` prior to constructing a node, which then automatically sets the node's location (in the `Obj` constructor). The parser's location is more precise than a node can obtain from its operands ilke in the example above - for example, in the expression `! foo`, the parser sets the location to start at `!` and end at the end of `foo`, whereas the example would have it only cover `foo`.

Regarding "easy to miss to keep consistent", there's an `ASSERT` in the ZAM compiler that complains if a node doesn't have _some_ sort of location associated with it (which is what happens for nodes constructed during optimization, unless they're associated with a predecessor like the logic in the PR laboriously does).

Reply:
> whereas the example would have it only cover foo.

Example here being the proposal I made. I think I understand now. An opposite situation currently: When an AddExpr is instantiated during parsing and we get down into `PromoteOps()`, the `ArithCoerceExpr`s created there for each of the ops will have a location covering the whole AddExpr, not the location of just the individual operands.

I have no good idea. Something that comes to mind is making location setting chain-able, but I'm not sure that just falls into the category of playing tricks to avoid the temporary/extra code.

Paging @bbannier for intuition.

In Expr.h (alternative names "WithLocationOf" or "AdoptLocationOf/From" or "CopyLocationOf").

```
+    ExprPtr UseLocationOf(const ExprPtr& other) {
+        SetLocationInfo(other->GetLocationInfo());
+        return ThisPtr();
+    }
+
```

```
--- a/src/Expr.cc
+++ b/src/Expr.cc
@@ -446,9 +446,7 @@ ExprPtr NameExpr::MakeLvalue() {
     if ( id->IsOption() && ! in_const_init )
         ExprError("option is not a modifiable lvalue");
 
-    auto r = make_intrusive<RefExpr>(ThisPtr());
-    r->SetLocation(this);
-    return r;
+    return make_intrusive<RefExpr>(ThisPtr())->UseLocationOf(ThisPtr());
 }
 
 void NameExpr::Assign(Frame* f, ValPtr v) {
@@ -1039,16 +1037,11 @@ void BinaryExpr::PromoteOps(TypeTag t) {
     if ( is_vec2 )
         bt2 = op2->GetType()->AsVectorType()->Yield()->Tag();
 
-    if ( bt1 != t ) {
-        auto orig_op1 = op1;
-        op1 = make_intrusive<ArithCoerceExpr>(op1, t);
-        op1->SetLocation(orig_op1);
-    }
-    if ( bt2 != t ) {
-        auto orig_op2 = op2;
-        op2 = make_intrusive<ArithCoerceExpr>(op2, t);
-        op2->SetLocation(orig_op2);
-    }
+    if ( bt1 != t )
+        op1 = make_intrusive<ArithCoerceExpr>(op1, t)->UseLocationOf(op1);
+
+    if ( bt2 != t )
+        op2 = make_intrusive<ArithCoerceExpr>(op2, t)->UseLocationOf(op2);
 }

Reply:
Benjamin suggested if anything to replace the class method and use a a free standing helper instead.
```
ExprPtr with_location(const ExprPtr&, Location *l);
ExprPtr with_location_of(const ExprPtr&, ExprPtr&l);

op1 = with_location(make_intrusive<ArithCoerceExpr>(op1, t), op1->GetLocationInfo())
op1 = with_location_of(make_intrusive<ArithCoerceExpr>(op1, t), op1)
```

Though neither this nor the verbose version are appealing or clear why really needed. Removing setting of an Obj's location based on parser state is likely the crux and would help disentangle (explicitly set locations `parse.y` on created nodes), but that's likely much bigger surgery.

@vpax, think I'd prefer the `with_location_of()` helper variant, but feel free to chime in if you think that's not actually helping. It's mostly just using a function argument as temporary with a bit more expressiveness :man_shrugging:

Reply:
`with_location_of()` WFM, I'll set about updating when I have some cycles